File: src/main/python/com/hazza/leetcode_by_python/linked_list/No86_Partition_List.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-
# Created by HazzaCheng on 2019-07-24
# https://leetcode.com/problems/partition-list/
from src.main.python.com.hazza.LeetCodeByPython.data_structure.ListNode import ListNode, makeListNode, isListNodeEqual


class Solution:
    def partition(self, head: ListNode, x: int) -> ListNode:
        if not head:
            return None
        smallStart, bigStart = ListNode(0), ListNode(0)
        smallPart, bigPart = smallStart, bigStart
        smallStart.next, bigStart.next = smallPart, bigPart

        while head:
            if head.val < x:
                smallPart.next = head
                smallPart = smallPart.next
            else:
                bigPart.next = head
                bigPart = bigPart.next
            head = head.next

        bigPart.next = None
        smallPart.next = bigStart.next

        return smallStart.next


if __name__ == '__main__':
    solution = Solution()

    head = makeListNode([1, 4, 3, 2, 5, 2])
    expected = makeListNode([1, 2, 2, 4, 3, 5])
    res = solution.partition(head, 3)
    assert isListNodeEqual(res, expected)

    head = makeListNode([1])
    expected = makeListNode([1])
    res = solution.partition(head, 2)
    assert isListNodeEqual(res, expected)

File: src/main/java/com/hazza/leetcode/linkedlist/No86PartitionList.java
package com.hazza.leetcode.linkedlist;

import com.hazza.leetcode.datastructure.ListNode;

/**
 * Created with IntelliJ IDEA.
 * Description: https://leetcode.com/problems/partition-list/description/
 * Author: HazzaCheng
 * Contact: hazzacheng@gmail.com
 * Date: 2017-09-23
 * Time: 2:39 PM
 */
public class No86PartitionList {
    public ListNode partition(ListNode head, int x) {
        ListNode start = new ListNode(0), bigStart = new ListNode(0);
        ListNode smallPart = start, bigPart = bigStart;
        start.next = smallPart;
        bigStart.next = bigPart;
        while (head != null) {
            if (head.val < x) smallPart = smallPart.next = head;
            else bigPart = bigPart.next = head;
            head = head.next;
        }
        bigPart.next = null;
        smallPart.next = bigStart.next;

        return start.next;
    }
}
